docs/design/1-overview.md

```diff
@@ -0,0 +1,479 @@
+The worker package will use the following interface structure (subject to refinement);
+
+```go
+const cgroupRoot = "/sys/fs/cgroup" // Adjust based on cgroup version
+
+// Command represents a job request with resource limits.
+type Command struct {
+	Name     string            // Command name (e.g., "ping") 
+	Args     []string          // Command arguments (e.g., ["-c", "4", "google.com"])
+	Env      map[string]string // Environment variable overrides
+	Timeout  time.Duration     // Max execution time before forceful stop
+	CPULimit string            // CPU limit (e.g., "0.5" = 50% of one core)
+	MemLimit string            // Memory limit (e.g., "500M")
+}
+
+// Job represents a running process with cgroup tracking.
+type Job struct {
+	ID     string             // Job identifier
+	Cmd    *exec.Cmd
+	Status *Status
+	cancel context.CancelFunc // Job cancellation
+	Cgroup string             // cgroup path
+	logBuf *LogBuffer         // Captures stdout/stderr for streaming
+}
+
+// Status represents the current state of a Job.
+type Status struct {
+	Pid      int  // Process identifier
+	ExitCode int  // Exit code (-1 if not exited)
+	Exited   bool // Whether the process has exited
+}
+
+// LogBuffer streams logs from a process.
+type LogBuffer struct {
+	mu       sync.Mutex
+	buffer   []string
+	subscribers map[chan string]struct{}
+}
+
+// Worker manages Linux jobs and their lifecycle.
+type Worker interface {
+	Start(command Command) (jobID string, err error)         // Starts a job
+	Stop(jobID string) error                                 // Stops a job
+	Query(jobID string) (status Status, err error)          // Queries a job's status
+	Stream(ctx context.Context, jobID string) (<-chan string, error) // Streams process output
+}
+```
+
+Key considerations for the worker package are to provide a comprehensive set of features for command control;
+
+* Override environment vars.
+* Gracefully handle job termination using `context.CancelFunc`
+* Concurrent-safe job management
+  * Will use `sync.Map` in the implementation (subject to refinement);
+```go
+// workerImpl is the concrete implementation of the Worker interface.
+type workerImpl struct {
+	jobs sync.Map     // Stores running jobs (jobID -> *Job)
+}
+  ```
+
+Additionally, future persistence can be attached without modifying this structure, as a subscriber. It allows for an arbitrary number (and type) of persistence services.
+
+## Process Execution Lifecycle (With cgroups)
+
+```mermaid
+sequenceDiagram
+    participant CLI as CLI User
+    participant gRPC as gRPC API
+    participant Worker as Job Worker
+    participant cgroup as Linux cgroup
+    participant Process as Linux Process
+
+    CLI->>gRPC: StartJob(command, args, env, limits)
+    gRPC->>Worker: Validate request, generate Job ID
+    Worker->>cgroup: Create cgroup (job-12345)
+    Worker->>cgroup: Set CPU/Memory limits
+    Worker->>Process: Start command inside cgroup
+    Process->>cgroup: Add PID to cgroup.procs
```

Comment:
Start the command inside which cgroup?
The process adds itself to the cgroup.procs file when it starts? This section is not clear to me. 

Can you please include more details?

Reply:
Thanks, I've amended the diagram. Essentially the process starts, then gets moved to the cgroup right away, written to the cgroup.procs file.